activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb

```diff
@@ -22,6 +22,7 @@ def accept(o)
 
       private
         def visit_AlterTable(o)
+          set_current_table(o.name)
```

Comment:
Would you let me know what is the purpose of setting table_name instance variable here?

Reply:
It is used to set the current table we create or alter and is needed here https://github.com/rails/rails/pull/45305/files#diff-5b1b920fc3bebbb1d4c144b302603e5003a1d9e5a493e3234bf764b9e26fbe30R181-R182 to be able to calculate the foreign key name based on the "from" and "to" table names in the `foreign_key_options` method. 

We can manually calculate the foreign key name and use something like
```ruby
# We need to get foreign key/ primary key columns here somehow,
# which is also done for us in `foreign_key_options`.
name = @conn.foreign_key_name(...)
options = @conn.foreign_key_options(nil, to_table, options.merge(name: name))
ForeignKeyDefinition.new(nil, to_table, options) <=== note that we set nil as table name
``` 

but this is ugly and more work that needed. Wdyt?

Reply:
Thanks for the update. I am not very familiar with design patterns, but I am not sure if it is good practice to set instance variables within a method that uses the visitor pattern with something like "set_something".

Reply:
I'm not super well versed in this design, but I don't think it's a big deal?

The visitor pattern for what I know of it, does down a tree to collect information and generate a result, so it doesn't shock me to keep some state when visiting.

But I think @tenderlove is much more well versed than me in Arel.